# Import Packages

We'll make use of the following packages:
- `numpy` is a package for scientific computing in python.
- `pandas` A powerful Python library for data manipulation and analysis.
- `seaborn` A data visualization library based on matplotlib.
- `scikit-learn` A comprehensive library for machine learning in Python.
- `kaggle` Using Kaggle API to download data.

In [2]:
import numpy as np
import pandas as pd
import kagglehub
import os
import seaborn as sns
import matplotlib.pyplot as plt


# Download Data

In [4]:
# Download latest version
path = kagglehub.dataset_download("taeefnajib/used-car-price-prediction-dataset")

print(os.listdir(path))

['used_cars.csv']


In [5]:
# Load the dataset
df = pd.read_csv(os.path.join(path, "used_cars.csv"))

# Data preprocessing

## Check data

In [6]:
df.head()

,brand,model,model_year,milage,fuel_type,engine,transmission,ext_col,int_col,accident,clean_title,price
0,Ford,Utility Police Interceptor Base,2013,"51,000 mi.",E85 Flex Fuel,300.0HP 3.7L V6 Cylinder Engine Flex Fuel Capa...,6-Speed A/T,Black,Black,At least 1 accident or damage reported,Yes,"$10,300"
1,Hyundai,Palisade SEL,2021,"34,742 mi.",Gasoline,3.8L V6 24V GDI DOHC,8-Speed Automatic,Moonlight Cloud,Gray,At least 1 accident or damage reported,Yes,"$38,005"
2,Lexus,RX 350 RX 350,2022,"22,372 mi.",Gasoline,3.5 Liter DOHC,Automatic,Blue,Black,None reported,NaN,"$54,598"
3,INFINITI,Q50 Hybrid Sport,2015,"88,900 mi.",Hybrid,354.0HP 3.5L V6 Cylinder Engine Gas/Electric H...,7-Speed A/T,Black,Black,None reported,Yes,"$15,500"
4,Audi,Q3 45 S line Premium Plus,2021,"9,835 mi.",Gasoline,2.0L I4 16V GDI DOHC Turbo,8-Speed Automatic,Glacier White Metallic,Black,None reported,NaN,"$34,999"


## 📝 Initial Data Glance – Observations & Next Steps

### Overview

- **Columns:**  
  The dataset contains **12 columns**: `brand`, `model`, `model_year`, `milage`, `fuel_type`, `engine`, `transmission`, `ext_col`, `int_col`, `accident`, `clean_title`, `price`.

- **Data Types:**  
  Most columns are **object type** (categorical or text). Data transformation will be required (label encoding, one-hot encoding, parsing text fields).

- **Brand & Model:**  
  - Some `model` values are duplicated or concatenated (e.g., `RX 350 RX 350`).
  - Will inspect and **clean/split/merge brand and model** to ensure unique, consistent values.

- **Feature Extraction:**  
  - Columns such as `engine` and `transmission` contain multiple details (e.g., horsepower, engine size, cylinder count, speed type) that can be **parsed into new features**.

- **Missing Values:**  
  - Nulls detected in several columns (e.g., `clean_title`).  
  - Will analyze missingness and apply appropriate imputation (mode, new category, or predictive imputation).

- **Target Variable:**  
  - The `price` column is a string with currency symbol and commas—needs to be cleaned and converted to numeric.

- **Formatting Issues:**  
  - Fields like `milage` and `price` contain units/symbols (e.g., "mi.", "$", ",")—will remove for numeric conversion.
  - `accident` and `clean_title` are categorical but may need binarization or mapping.

---

### Next Steps

- Clean and standardize all categorical and text fields.
- Parse and extract features from `engine` and `transmission` columns.
- Handle missing values with suitable imputation strategies.
- Convert `price` and `milage` to numeric types.
- Ensure brand/model consistency for analysis and modeling.